modules/basegl/src/processors/instancerenderer.cpp

```diff
@@ -0,0 +1,394 @@
+                if (prop == property) continue;
+
+                auto comp = dynamic_cast<CompositeProperty*>(prop);
+                auto uniform =
+                    dynamic_cast<StringProperty*>(comp->getPropertyByIdentifier("uniform"));
+                if (uniform->get() == id) return false;
+            }
+            return true;
+        },
+        "");
+    uniform->set(uniqueID);
+
+    const auto& id = property->getIdentifier();
+    switch (components->get()) {
+        case 1:
+            vecPorts_.push_back(detail::createDynPortManager<float>(this, id, uniform));
+            break;
+        case 2:
+            vecPorts_.push_back(detail::createDynPortManager<vec2>(this, id, uniform));
+            break;
+        case 3:
+            vecPorts_.push_back(detail::createDynPortManager<vec3>(this, id, uniform));
+            break;
+        case 4:
+            vecPorts_.push_back(detail::createDynPortManager<vec4>(this, id, uniform));
+            break;
+    }
+
+    invalidate(InvalidationLevel::InvalidResources);
+}
+
+void InstanceRenderer::onWillRemoveProperty(Property* property, size_t index) {
+    vecPorts_.erase(vecPorts_.begin() + index);
+
+    invalidate(InvalidationLevel::InvalidResources);
+}
+
+void InstanceRenderer::initializeResources() {
+    utilgl::addShaderDefines(shader_, lightingProperty_);
+
+    auto* vso = shader_.getVertexShaderObject();
+    vso->clearSegments();
+
+    vso->addSegment(
+        ShaderSegment{irplaceholder::setupVert, setupVert_.getIdentifier(), setupVert_.get()});
+
+    std::for_each(vecPorts_.begin(), vecPorts_.end(), [&](auto& port) { port.addUniform(*vso); });
+
+    size_t prio = 100;
+    for (auto& transform : transforms_) {
+        vso->addSegment(ShaderSegment{
+            irplaceholder::transforms, transform.getIdentifier(),
+            fmt::format("{} = {};", transform.getIdentifier(), transform.get()), prio});
+        prio += 100;
+    }
+
+    shader_.build();
+}
+
+void InstanceRenderer::process() {
+    utilgl::activateTargetAndClearOrCopySource(outport_, imageInport_);
+
+    if (vecPorts_.empty()) return;
+
+    shader_.activate();
+
+    utilgl::GlBoolState depthTest(GL_DEPTH_TEST, true);
+    utilgl::BlendModeState blendModeStateGL(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+
+    utilgl::setUniforms(shader_, camera_, lightingProperty_);
+
+    const auto& mesh = *inport_.getData();
+    MeshDrawerGL::DrawObject drawer{mesh.getRepresentation<MeshGL>(), mesh.getDefaultMeshInfo()};
+    utilgl::setShaderUniforms(shader_, mesh, "geometry");
+
+    auto minIt = std::min_element(vecPorts_.begin(), vecPorts_.end(),
+                                  [](const auto& a, const auto& b) { return a.size() < b.size(); });
+    const size_t min = minIt->size();
+
+    for (size_t i = 0; i < min; ++i) {
+        std::for_each(vecPorts_.begin(), vecPorts_.end(),
+                      [&](auto& port) { port.set(shader_, i); });
+
+        drawer.draw();
```

Comment:
No glDrawArrayInstanced?

Reply:
should be possible to change into that I think, have not had the time/need for it so far...

Reply:
Might go either way. This way it supports generic meshes. And since the uniforms have to be set before each draw call anyway...